File: src/main/java/org/starlisp/core/LispChar.java
package org.starlisp.core;

final class LispChar extends LispObject {
  public final char ch;

  public LispChar(char ch) {
    this.ch = ch;
  }

  public int hashCode() {
    return Character.valueOf(ch).hashCode();
  }

  public boolean equals(Object obj) {
    return (obj instanceof LispChar) ? ((LispChar) obj).ch == ch : false;
  }

  public String toString() {
    return "#\\" + ch;
  }
}

File: src/main/scala/org/starlisp/core/LispChar.scala
package org.starlisp.core

import collection.mutable

object LispChar {
  val lowCache = {
    val lc = new Array[LispChar](256)
    (0 until 256).foreach{idx => lc(idx) = new LispChar(idx.asInstanceOf[Char])}
    lc
  }

  val cache = new mutable.HashMap[Char, LispChar]()
  def create(ch: Char) = {
    if (ch < 256) {
      lowCache(ch)
    } else {
      cache.getOrElseUpdate(ch, new LispChar(ch))
    }
  }
}

class LispChar(val ch: Char = 0) extends LispObject {
  override def hashCode = ch.hashCode
  override def equals(obj: Any) = obj.isInstanceOf[LispChar] && (obj.asInstanceOf[LispChar]).ch == ch
  override def toString = "#\\%c".format(ch)
}